src/lib/update-lock.ts

```diff
@@ -76,6 +77,28 @@ async function dispose(
 	}
 }
 
+/**
+ * Composition step used by Supervisor compose module.
+ * Take all locks for an appId | appUuid, creating directories if they don't exist.
+ */
+export async function takeLock(appId: number, services: string[]) {
+	const release = await takeGlobalLockRW(appId);
+	try {
+		// Filter out services that already have Supervisor-taken locks.
+		// This needs to be done after taking the appId write lock to avoid
+		// race conditions with locking.
+		const servicesWithoutLock = services.filter(
+			(svc) => !getServicesLockedByAppId().isLocked(appId, svc),
+		);
+		for (const service of servicesWithoutLock) {
+			await mkdirp(pathOnRoot(lockPath(appId, service)));
+			await lockService(appId, service);
```

Comment:
I think using `lockService` is a problem here. Lock service is not "all or nothing", it could take `update.lock` but fail on `resin-update.lock`. That was solved on the [lock](https://github.com/balena-os/balena-supervisor/blob/master/src/lib/update-lock.ts#L94) function by cleaning up locks on exception, but that is not the case here

Reply:
Could `takeLock` perhaps call `lock` and not receive an array of services?

Reply:
> Could `takeLock` perhaps call `lock` and not receive an array of services?

I'm not sure this would work, as update-lock does not know which services should be locked under an appId. If the service directories under an appId are always assumed to be there by the update-lock module, this wouldn't be an issue because update-lock could be just read a list of services under an appId and proceed with that as the services to be locked.

However, `takeLock` can be called for a service in the target state that hasn't started up, meaning the `/tmp` bind mounts for locking haven't necessarily been created yet by the Supervisor. Therefore `takeLock` needs an array of services to know which directories to create

Reply:
I don't think `lock` would work as it only temporarily locks an app to perform an action; A `dispose` which removes all locks is performed in the same `lock` function which removes all locks regardless of whether the lock errored or not. Is that what you're referring to by lock cleanup? I don't see any other places where the lock may be cleaned up.

Reply:
I see your point though about succeeding on one of `(resin-)update.lock` but failing on the other; if `lockService` fails in this case, both lockfiles need to be removed which currently `lockService` doesn't do. This PR which touches `lockService` directly has exposed the flaws in `lockService` when it's not immediately paired with a `dispose` as is the case in `lock`. Thus, I think we can patch `lockService` to remove both lockfiles instead, what do you think?

As a side note, I mentioned that one of the 2 lockfiles failing to lock means the service should not count as locked, but if only one of the two files exists already, the service should count as locked. This is because `resin-updates.lock` is a legacy file, and while we should still support it, we shouldn't make the success of lock-taking contingent on its existence.

Reply:
> Therefore takeLock needs an array of services to know which directories to create

You could have an earlier step to create the target directories manually and `takeLock` with just the appId can try to take all of the locks for the app at once. If one fails, it needs to dispose the locks and throw so it can try again

Alternatively, `takeLock` may receive a list of services (which is assumed to contain all services of the app) and first create the missing directories and then try to take all the locks. Again, if something fails, it throws and the state engine needs to run again

> I don't think lock would work as it only temporarily locks an app to perform an action

You are right, I forgot `lock` receives a function. The point stands, `lockService` could fail in the middle of taking the two lockfiles for a service which could be a problem (it won't be if we require both files to exist with `nobody` to consider the lock taken). As with lock, I think we need to figure out a way to take all the locks or fail as I pointed out before to avoid potential deadlock with the user app

Reply:
Sounds good, I think we could change takeLock steps' implementation to catch any locking failures, remove any taken locks, then throw the error again to be picked up by the state engine so it can retry

Reply:
I added the lockfile cleanup on lock failure in a separate commit: 48e869104620f6c8447e5abbe2af9ee40bce9605